internal/manifest/oci_index.go

```diff
@@ -64,22 +74,52 @@ func (index *OCI1IndexPublic) Instance(instanceDigest digest.Digest) (ListUpdate
+		editInstances = append(editInstances, editInstance)
 	}
-	for i := range updates {
-		if err := updates[i].Digest.Validate(); err != nil {
-			return fmt.Errorf("update %d of %d passed to OCI1Index.UpdateInstances contained an invalid digest: %w", i+1, len(updates), err)
-		}
-		index.Manifests[i].Digest = updates[i].Digest
-		if updates[i].Size < 0 {
-			return fmt.Errorf("update %d of %d passed to OCI1Index.UpdateInstances had an invalid size (%d)", i+1, len(updates), updates[i].Size)
-		}
-		index.Manifests[i].Size = updates[i].Size
-		if updates[i].MediaType == "" {
-			return fmt.Errorf("update %d of %d passed to OCI1Index.UpdateInstances had no media type (was %q)", i+1, len(updates), index.Manifests[i].MediaType)
+	return index.EditInstances(editInstances)
+}
+
+func (index *OCI1IndexPublic) EditInstances(editInstances []ListEdit) error {
+	addedEntries := []imgspecv1.Descriptor{}
+	for i, editInstance := range editInstances {
+		switch editInstance.ListOperation {
+		case listOpUpdate:
+			if err := editInstance.UpdateOldDigest.Validate(); err != nil {
+				return fmt.Errorf("OCI1Index.EditInstances: Attempting to update %s which is an invalid digest: %w", editInstance.UpdateOldDigest, err)
+			}
+			if err := editInstance.UpdateDigest.Validate(); err != nil {
+				return fmt.Errorf("OCI1Index.EditInstances: Modified digest %s is an invalid digest: %w", editInstance.UpdateDigest, err)
+			}
+			instanceToBeModifiedIdx := index.descriptorIndex(editInstance.UpdateOldDigest)
+			if instanceToBeModifiedIdx < 0 {
+				return fmt.Errorf("OCI1Index.EditInstances: Attempting to update %s which is an invalid digest", editInstance.UpdateOldDigest)
+			}
+			index.Manifests[instanceToBeModifiedIdx].Digest = editInstance.UpdateDigest
+			if editInstance.UpdateSize < 0 {
+				return fmt.Errorf("update %d of %d passed to OCI1Index.UpdateInstances had an invalid size (%d)", i+1, len(editInstances), editInstance.UpdateSize)
+			}
+			index.Manifests[instanceToBeModifiedIdx].Size = editInstance.UpdateSize
+			if editInstance.UpdateMediaType == "" {
+				return fmt.Errorf("update %d of %d passed to OCI1Index.UpdateInstances had no media type (was %q)", i+1, len(editInstances), index.Manifests[i].MediaType)
+			}
+			index.Manifests[instanceToBeModifiedIdx].MediaType = editInstance.UpdateMediaType
+		case listOpAdd:
+			addedEntries = append(addedEntries, imgspecv1.Descriptor{
+				MediaType:   editInstance.AddMediaType,
+				Size:        editInstance.AddSize,
+				Digest:      editInstance.AddDigest,
+				Platform:    editInstance.AddPlatform,
+				Annotations: editInstance.AddAnnotations})
+			index.Manifests = append(index.Manifests, addedEntries...)
+			sort.SliceStable(index.Manifests, func(i, j int) bool {
+				return !instanceIsZstd(index.Manifests[i]) && instanceIsZstd(index.Manifests[j])
+			})
```

Comment:
Sorting on every add does work, but it ends up being O(N^2 long N) instead of O(N log N), unnecessarily.

Just collect `addedEntries` while processing edits, and at the end, if `len(addedEntries) != 0`, actually append and sort once.

(Adding `addedEntries` at the very end also has the benefit of failing on a request to update a just-added entry, a little bit of extra validation.)

Reply:
done.